File: 396_Rotate_Function.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution{
public:
    int maxRotateFunction(vector<int>& A){
        if (A.empty()) return 0;
        int n = A.size(); int sum = 0;
        vector<int> f(n);

        for (int s : A) sum += s;
        for (int i = 0; i < n; ++i) f[0] += i * A[i];

        for (int i = 1; i < n; ++i) {
            f[i] = f[i - 1] + sum - n * (long long)A[n - i];
        }

        auto res = max_element(f.begin(), f.end());
        return *res;
    }
}; Solution solution;


int main(){
    vector<int> A{4, 3, 2, 6};
    auto res = solution.maxRotateFunction(A);
    
    cout << res << endl;

    return 0;
}

File: 396_Rotate_Function.py
class Solution(object):
    def maxRotateFunction(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        if not A:
            return 0
            
        s, n = sum(A), len(A)
        f = [0] * n
        for i, num in enumerate(A):
            f[0] += i * A[i]

        mx = f[0]
        for i in range(1, n):
            f[i] = f[i - 1] + s - n * A[n - i]
            mx = max(mx, f[i])
        return mx

solution = Solution()
print(solution.maxRotateFunction([4, 3, 2, 6]))